test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -230,25 +271,61 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
 		dockerMachinePool.Spec.ProviderID = getDockerMachinePoolProviderID(cluster.Name, dockerMachinePool.Name)
```

Comment:
If CAPD actually runs kubelet (I think it does?) and it has a cloud-init context from which the CAPD provider id can be constructed (I have no idea), the most robust way to do this is to pass provider-id to kubelet via its config, e.g.: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/739fbe36470aeacb9a6dbd843c54117b9fb09f4d/templates/cluster-template.yaml#L23. In this way kubelet sets providerID on the Node when it is initially created.

Reply:
Interesting - CAPD does run kubelet, but it doesn't run Cloud Init. Does that method of setting the providerID conflict with an external cloud provider?

Reply:
@killianmuldoon On the contrary: it solves a problem for the external cloud provider because it means it doesn't have to try to heuristically match a node to a cloud resource in order to set the providerID, it can just use the one that's already set.

Reply:
Interesting - CAPD has a cloud-init adapter where it emulates the essential parts of cloud-init for tests, we could  probably implement setting the provider ID there.

We still need CAPD to do patching for network addresses for dualstack unfortunately, but that could be solved in an upcoming Kubernetes release. Would also help in getting rid of the `[SHOULD NOT HAPPEN]` logs from our dualstack tests :sweat_smile:

Reply:
@killianmuldoon Gotcha, is there an issue tracking dual stack e2e with MPs?

Reply:
It's tracked as part of https://github.com/kubernetes-sigs/cluster-api/issues/5991 for now - under E2E tests.